client/fetch_tagged_results_accumulator.go

```diff
@@ -0,0 +1,352 @@
+		}
+
+		if resultErr == nil {
+			shardResult.success++
+		} else {
+			shardResult.errors++
+		}
+
+		shardResult.done = readConsistencyAchieved(accum.consistencyLevel, accum.majority,
+			int(shardResult.enqueued), int(shardResult.success))
+
+		if shardResult.done {
+			accum.numShardsPending--
+		}
+
+		// update value in slice
+		accum.shardResults[shardID] = shardResult
+	}
+
+	// success case, sufficient responses for each shard
+	if accum.numShardsPending == 0 {
+		doneAccumulating := true
+		return doneAccumulating, nil
+	}
+
+	// failure case - we've received all responses but still weren't able to satisfy
+	// all shards, so we need to fail
+	if accum.numHostsPending == 0 && accum.numShardsPending != 0 {
+		doneAccumulating := true
+		return doneAccumulating, fmt.Errorf(
+			"unable to satisfy consistency requirements for %d shards [ err = %s ]",
+			accum.numShardsPending, accum.errors.Error())
+	}
+
+	doneAccumulating := false
+	return doneAccumulating, nil
+}
+
+func (accum *fetchTaggedResultAccumulator) Clear() {
+	accum.shardResults = accum.shardResults[:0]
+	for i := range accum.responses {
+		accum.responses[i] = nil
+	}
+	accum.responses = accum.responses[:0]
+	for i := range accum.errors {
+		accum.errors[i] = nil
+	}
+	accum.errors = accum.errors[:0]
+	accum.consistencyLevel = topology.ReadConsistencyLevelNone
+	accum.majority, accum.numHostsPending, accum.numShardsPending = 0, 0, 0
+	accum.startTime, accum.endTime = time.Time{}, time.Time{}
+	accum.topoMap = nil
+}
+
+func (accum *fetchTaggedResultAccumulator) Reset(
+	startTime time.Time,
+	endTime time.Time,
+	topoMap topology.Map,
+	majority int,
+	consistencyLevel topology.ReadConsistencyLevel,
+) {
+	accum.exhaustive = true
+	accum.startTime = startTime
+	accum.endTime = endTime
+	accum.topoMap = topoMap
+	accum.majority = majority
+	accum.consistencyLevel = consistencyLevel
+	accum.numHostsPending = int32(topoMap.HostsLen())
+	accum.numShardsPending = int32(len(topoMap.ShardSet().All()))
+
+	// expand shardResults as much as necessary
+	targetLen := 1 + int(topoMap.ShardSet().Max())
+	accum.shardResults = fetchTaggedShardConsistencyResults(accum.shardResults).initialize(targetLen)
+	// initialize shardResults based on current topology
+	for _, hss := range topoMap.HostShardSets() {
+		for _, hShard := range hss.ShardSet().All() {
+			id := int(hShard.ID())
+			accum.shardResults[id].enqueued++
+		}
+	}
+}
+
+func (accum *fetchTaggedResultAccumulator) sliceResponsesAsSeriesIter(
+	pools fetchTaggedPools,
+	elems fetchTaggedIDResults,
+) encoding.SeriesIterator {
+	numElems := len(elems)
+	iters := pools.IteratorArray().Get(numElems)[:numElems]
+	for idx, elem := range elems {
+		slicesIter := pools.ReaderSliceOfSlicesIterator().Get()
+		slicesIter.Reset(elem.Segments)
+		multiIter := pools.MultiReaderIterator().Get()
+		multiIter.ResetSliceOfSlices(slicesIter)
+		iters[idx] = multiIter
+	}
+
+	elem := elems[0] // pick the first element as they're all identical
```

Comment:
Can you explain the guarantee that this slice wont be empty

Reply:
This method `sliceResponsesAsSeriesIter` is called from within a lambda provided to `func (results fetchTaggedIDResults) forEach(fn forEachFetchTaggedIDFn)` which makes that guarantee. 

i.e. forEach is only called on a group, impossible to have a group with 0 members.